### PR TITLE
fix: don't assume files to have CR characters on Windows

### PR DESCRIPTION
**Pull request by shinnn**
The current code assumes that all lines of a JavaScript file end with CR+LF on Windows. However, for example, if the `eol` attribute of the project's .gitattributes is [`lf`](https://git-scm.com/docs/gitattributes#Documentation/gitattributes.txt-Settostringvaluelf), files don't have CR characters even on Windows:

```console
$ node -p "fs.readFileSync('1.js')" # when .gitattribute is "* text=auto"
<Buffer 31 0d 0a>

$ node -p "fs.readFileSync('1.js')" # when .gitattribute is "* text=auto eol=lf"
<Buffer 31 0a>
```

Because of this assumption, currently v8-to-istanbul cannot calculate line coverage correctly especially on Windows.

This patch solves the problem above.

### CODE REVIEW

File: lib/script.js [hunk 1]
```diff
@@ -28,14 +27,15 @@ module.exports = class CovScript {
   }
   _buildLines (source, lines, shebangLength) {
     let position = 0
-    const separator = isWindows ? '\r\n' : '\n'
-    ;(source.trim()).split(separator).forEach((lineStr, i) => {
-      this.eof = position + lineStr.length
+    for (const [i, lineStr] of source.trim().split(/(?<=\r?\n)/u).entries()) {
+      const matchedNewLineChar = lineStr.match(/\r?\n$/u)
+      const newLineLength = matchedNewLineChar ? matchedNewLineChar[0].length : 0
```
Comment (shinnn) [inline lib/script.js]: > ```js
> matchedNewLineChar ? matchedNewLineChar[0].length : 0
> ```

Files don't always end with a newline.

File: lib/script.js [hunk 2]
```diff
@@ -28,14 +27,15 @@ module.exports = class CovScript {
   }
   _buildLines (source, lines, shebangLength) {
     let position = 0
-    const separator = isWindows ? '\r\n' : '\n'
-    ;(source.trim()).split(separator).forEach((lineStr, i) => {
-      this.eof = position + lineStr.length
+    for (const [i, lineStr] of source.trim().split(/(?<=\r?\n)/u).entries()) {
```
Comment (shinnn) [inline lib/script.js]: Even if a file contains both LF and CR+LF, `/(?<=\r?\n)/u` splits contents as expected.

```javascript
'First\nSecond\r\nThird'.split(/(?<=\r?\n)/u)
//=> ['First\n', 'Second\r\n', 'Third']
```

Comment (shinnn) [reply lib/script.js]: As this library doesn't support legacy Node.js versions, we can safely use lookbehind assertions. https://node.green/#ES2018-features--RegExp-Lookbehind-Assertions

**Comment by bcoe** [comment]
@shinnn this appears to drop coverage significantly any thoughts? are you able to run coverage locally using the latest `c8`?

**Comment by bcoe** [comment]
btw, `coverage` is really cool :+1: will happily continue working with you to make `c8` better.

**Comment by shinnn** [comment]
Local coverage reports on the `master` branch:

```console
$ c8 --reporter=html --reporter=text tap --no-esm test/*.js

test/script.js ........................................ 8/8
total ................................................. 8/8

  8 passing (298.291ms)

  ok
--------------------|----------|----------|----------|----------|-------------------|
File                |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
--------------------|----------|----------|----------|----------|-------------------|
All files           |    88.75 |    95.74 |      100 |    88.75 |                   |
 v8-to-istanbul     |    33.33 |      100 |      100 |    33.33 |                   |
  index.js          |    33.33 |      100 |      100 |    33.33 |           3,4,5,6 |
 v8-to-istanbul/lib |    90.17 |    95.56 |      100 |    90.17 |                   |
  branch.js         |    78.57 |      100 |      100 |    78.57 | 23,24,25,26,27,28 |
  function.js       |    75.86 |      100 |      100 |    75.86 |... 25,26,27,28,29 |
  line.js           |    71.43 |      100 |      100 |    71.43 | 16,17,18,19,20,21 |
  script.js         |    97.44 |    94.44 |      100 |    97.44 |   153,154,155,156 |
--------------------|----------|----------|----------|----------|-------------------|
```

On the branch of this PR:

```console
$ c8 --reporter=html --reporter=text tap --no-esm test/*.js

test/script.js ...................................... 10/10
total ............................................... 10/10

  10 passing (275.907ms)

  ok
--------------------|----------|----------|----------|----------|-------------------|
File                |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
--------------------|----------|----------|----------|----------|-------------------|
All files           |    88.75 |    97.92 |      100 |    88.75 |                   |
 v8-to-istanbul     |    33.33 |      100 |      100 |    33.33 |                   |
  index.js          |    33.33 |      100 |      100 |    33.33 |           3,4,5,6 |
 v8-to-istanbul/lib |    90.17 |    97.83 |      100 |    90.17 |                   |
  branch.js         |    78.57 |      100 |      100 |    78.57 | 23,24,25,26,27,28 |
  function.js       |    75.86 |      100 |      100 |    75.86 |... 25,26,27,28,29 |
  line.js           |    71.43 |      100 |      100 |    71.43 | 16,17,18,19,20,21 |
  script.js         |    97.44 |     97.3 |      100 |    97.44 |   153,154,155,156 |
--------------------|----------|----------|----------|----------|-------------------|
```

I'd rather doubt almost 100% coverage on the latest Travis CI build. https://travis-ci.org/istanbuljs/v8-to-istanbul/builds/516738314#L498-L509

**Comment by bcoe** [comment]
@shinnn do me a favor and rebase with `master`, I believe coverage should be higher now that I've upgraded the version of c8 that tests the library.

**Comment by shinnn** [comment]
> Overall coverage increased (+**2.009%**) to **99.647%**

Done!

**Comment by bcoe** [comment]
published in `v8-to-istanbul@2.0.5` thanks for the awesome contribution :+1:

**Comment by shinnn** [comment]
@bcoe No problem. Thanks for adding me to this project and c8 as a collaborator!

**Pull request merged**